### ISSUE TITLE
Code Review 0.4.2..0.4.3-rc3

### ISSUE THREAD

**Issue by jbenet** [ipfs/go-ipfs#3138]
I went through and code reviewed all the PRs, commits, and merges from `0.4.2..0.4.3-rc3`.

![](https://cdn.stepfeed.com/uploads/2015/10/13151456/20151013-so-productive.gif)

First, a huge thanks to everyone. this is a huge release, with tremendous stuff. Huge shout out to everyone who worked on it, sp @whyrusleeping @Kubuxu @kevina @lgierth @RichardLitt @chriscool and all those listed here https://github.com/ipfs/go-ipfs/blob/master/CHANGELOG.md#043-rc3---2016-08-09 congrats, this is a great release with a ton of bug fixes, stability improvements, and new features. 🎉 🎆 🎊 🎈 🎉 

![](http://gifsec.com/wp-content/uploads/GIF/2014/11/Happy-Cat-GIF.gif?gs=a)

![](https://media.giphy.com/media/FYx64DDl2ElWw/giphy.gif)
### On this Code Review

My code review was for the purpose of finding any remaining problems before the `0.4.3` release. I read all of the code. I didn't look at everything super carefully; i focused on critical sections, UX, API changes, and security issues. I've left a bunch of comments all over the place (Merges, PRs, commits).

I took specific notes of things to address. The "MUST be addressed by `0.4.3`" ones should have a commit or PR link to a _merged_ thing (edit this issue's list to add it next to the line item when you check it) before release. It's not that much, but it's also substantial. When done, please add the link like this at the end of the line: 

```
- [ ] <line item> -- resolved in <link-to-merged-PR>
```

By design, this whole list is _negative_, as the goal was identifying problems. Meaning, i only highlighted all the stuff that _needs to change_. I looked at every commit, so a lacking comment here means that those changes implicitly LGTM, and I'm fine with releasing them. So thank you for all that! 👍 😄 🎉 🎉 

It's up to @whyrusleeping whether to do these changes on top of `master` or on top of `0.4.3-rc3`. If we want to roll the latest changes (`0.4.3-rc3..master`) into `0.4.3` (instead of `0.4.4`), then i will need to code review that section too. I will get started on that anyway, and should have it done by monday eve.
### MUST happen before 0.4.3 release
- [x] failed to ipfs-update install v0.4.3-rc3 -- fs-repo-migrations https://github.com/ipfs/fs-repo-migrations/issues/45
- [ ] change for ipfs refs local in the api. needs better changelog entry and option? https://github.com/ipfs/go-ipfs/pull/2812#issuecomment-242672559
- [x] needs a specific test -- https://github.com/ipfs/go-ipfs/pull/2778 --- resolved in https://github.com/ipfs/go-ipfs/pull/3142
- [x] license question -- https://github.com/ipfs/go-ipfs/pull/2822/files#r76513373
- [x] tests from the gateway fail need to be back @lgierth
  - [ ] actually, i thought that PR was entirely reverted... how are we sure there aren't other issues? -- A: fixing the refactored (broken) code is a net win over reverting it all, resolved in https://github.com/ipfs/go-ipfs/pull/2874#issuecomment-243615655
  - [ ] no longer forbidden bug? -- A: nope, resolved in https://github.com/ipfs/go-ipfs/pull/2874#discussion_r76897912
  - [x] #3159 brings back --unrestricted-api bash auto completion
- [x] best effort file api pin needs testing https://github.com/ipfs/go-ipfs/pull/2872/files#r76513709 --- needs review #3151 
  - [ ] best effort file api pin is an important product + protocol decision i never knew about. not great, clear these with me
  - [ ] best effort pins implemented poorly https://github.com/ipfs/go-ipfs/pull/2872#issuecomment-242908207
- [x] need to revert symlink resolve -- https://github.com/ipfs/go-ipfs/pull/2897 --- resolved in https://github.com/ipfs/go-ipfs/pull/3135
  - [x] revert this too -- https://github.com/ipfs/go-ipfs/pull/2926 --- already reverted in https://github.com/ipfs/go-ipfs/pull/2952
- [x] auto raise fd limit needs better testing https://github.com/ipfs/go-ipfs/pull/2891/files#r76513912 --- needs review #3149
- [x] CLI stdin breaking changes? -- https://github.com/ipfs/go-ipfs/pull/2822#issuecomment-242906605
  - [x] cli stdin redux, not reverted? -- https://github.com/ipfs/go-ipfs/pull/2902/files#r76514113 (https://github.com/ipfs/go-ipfs/pull/3144#issuecomment-243658915)
- [x] proper repo verify testing -- https://github.com/ipfs/go-ipfs/pull/2924/files#r76514755
- [x] PrivKey rm security harden -- https://github.com/ipfs/go-ipfs/pull/2957/files#r76539226
  - [ ] insecure measure -- https://github.com/ipfs/go-ipfs/pull/2957/files#r76539324
  - [ ] lower case test case? -- https://github.com/ipfs/go-ipfs/pull/2957/files#r76539414
  - [x] Merged #3141 
- [x] auto-run migration doesnt have latest migrations, right? -- https://github.com/ipfs/go-ipfs/pull/2939/files
  - [x] better errors in auto-install fail @lgierth -- https://github.com/ipfs/go-ipfs/pull/2939/files#r76539747
    - RFCR #3158
  - [x] more errors @lgierth -- https://github.com/ipfs/go-ipfs/pull/2939/files#r76539800
    - RFCR #3158
### Important but postponed to 0.4.4
- [ ] blockstore.{RuntimeHashing -> HashOnRead} -- https://github.com/ipfs/go-ipfs/pull/2904/files#r76514037
  ### Other
- [ ] ipfs swarm peers -v may output a 0 latency -- https://github.com/ipfs/go-ipfs/issues/3125
- [ ] 2482dc8 merged two commits, one which undid the other. Probably best to ask author to rebase and remove the "undone" code.
- [ ] ipfs files cp/write erros and whatnot -- https://github.com/ipfs/go-ipfs/issues/3126
- [ ] ipfs files stat buggy default output -- https://github.com/ipfs/go-ipfs/issues/3127
- [ ] please rebase PRs before merging pls. it makes long term CR easier. (particularly @kubuxu)
- [ ] undo #2657 language and progress bar -- https://github.com/ipfs/go-ipfs/pull/2657#issuecomment-242660570
- [ ] offline mode PR feedback -- https://github.com/ipfs/go-ipfs/pull/2696/files#r76381690
- [ ] #2634 Adder error prone? -- https://github.com/ipfs/go-ipfs/pull/2792/files#r76384826
- [ ] needs squashing commits. non passing were merged https://github.com/ipfs/go-ipfs/pull/2598
- [ ] regex fix https://github.com/ipfs/go-ipfs/pull/2598/files#r76388121
- [ ] some improvements to dht messageSender code in https://github.com/ipfs/go-ipfs/pull/2817/files#r76512847
- [x] thank you for adding tests. PRs with only tests are great.
- [ ] dup tests into graphs? https://github.com/ipfs/go-ipfs/pull/2798
- [x] not super clear why -- https://github.com/ipfs/go-ipfs/pull/2813
- [ ] clarify RoutingTable protects -- https://github.com/ipfs/go-ipfs/pull/2841/files#r76513202
- [x] yay for all the gx-ification!
- [ ] makefile change buried in another PR https://github.com/ipfs/go-ipfs/pull/2855/files#r76513294
- [ ] SO_REUSEPORT totally disabled in go tests, scary -- https://github.com/ipfs/go-ipfs/pull/2855#issuecomment-242906322
- [ ] codify API changes into commits -- https://github.com/ipfs/go-ipfs/pull/2867
- [ ] come up with guidelines for timeouts -- https://github.com/ipfs/go-ipfs/pull/2863
- [ ] fd limits question https://github.com/ipfs/go-ipfs/pull/2884/files#r76513661
- [ ] copyright notice needs fixing https://github.com/ipfs/go-ipfs/pull/2883/files#r76513693
- [ ] autogenerated synopsis, short usage? -- https://github.com/ipfs/go-ipfs/pull/2785#issuecomment-242910185
- [ ] abstraction leak, ls resolve flag -- https://github.com/ipfs/go-ipfs/pull/2824/files#r76514362
- [x] very nice write up + thinking through first -- https://github.com/ipfs/go-ipfs/issues/2601#issuecomment-227340622
- [ ] benchmark graphs for blockstore arc cache -- https://github.com/ipfs/go-ipfs/pull/2885#issuecomment-242911519
- [ ] some CR notes on https://github.com/ipfs/go-ipfs/pull/2953/files
- [ ] some CR notes on https://github.com/ipfs/go-ipfs/pull/2960/files#r76539147
- [ ] some comments on - https://github.com/ipfs/go-ipfs/pull/2974
- [ ] test migration bin names https://github.com/ipfs/go-ipfs/pull/3021
- [ ] why is this lock here? https://github.com/ipfs/go-ipfs/pull/3026/files#r76540554
- [ ] test these changes? https://github.com/ipfs/go-ipfs/pull/3023#issuecomment-243003995
- [ ] comment these things -- https://github.com/ipfs/go-ipfs/pull/3036/files#r76540668

**Comment by Kubuxu** [comment]
About rebasing PRs before merging, I can't do much in this manner, it is Jeromy who merges them, he can ping me to rebase them (he does in case of conflict) or rebase them himself (my signature won't match then).

**Comment by whyrusleeping** [comment]
@jbenet thanks for all the review! The strategy will be to branch off of the 0.4.3-rc3 tag and merge everything to there. Then merge that into master once the release happens.

**Comment by Kubuxu** [comment]
I already created `version/0.4.3-rc4` which is based off the tag and there are two PRs created to it: https://github.com/ipfs/go-ipfs/pull/3134 https://github.com/ipfs/go-ipfs/pull/3135

**Comment by lgierth** [comment]
Another lesson from the 0.4.3 cycle is that we should only start releasing RC builds once we're absolutely positive that this is what we want to release. We're close to a 1.5 months RC cycle.

**Comment by Kubuxu** [comment]
@jbenet Is ` blockstore.{RuntimeHashing -> HashOnRead}` neccesary for 0.4.3, it would be best to keep the changeset minimal in the 0.4.3 branch to go into merge conflict hell.

As it is small refactor (and master has already some changes with it), it would be better to keep it on master. Sounds good?


@lgierth I don't know why having long RC cycles is bad thing, it only show that our QA game needs improvement but including community in throughout testing is a good thing.

**Comment by jbenet** [comment]
> @jbenet Is `blockstore.{RuntimeHashing -> HashOnRead}` neccesary for `0.4.3`, it would be best to keep the changeset minimal in the `0.4.3` branch to go into merge conflict hell.

@Kubuxu it's an interface. our releases are not only binaries, they're also the code. It's pretty minimal, but okay.

**Comment by lgierth** [comment]
> @lgierth I don't know why having long RC cycles is bad thing, it only show that our QA game needs improvement but including community in throughout testing is a good thing.

Sure -- I just mean that if we know that @jbenet is still to review it, and is likely going to have lots of feedback, it's IMO not yet a release candidate. A release candidate is "almost exactly this is going to be the release, unless there are bad bugs or regressions".

**Comment by jbenet** [comment]
Agree with @lgierth

**Comment by kevina** [comment]
@jbenet (also @whyrusleeping @Kubuxu) concerning "best effort" pins, let me know what I need to do.  At the moment this seams to be due to a misunderstanding, see #2698.

**Comment by RichardLitt** [comment]
Thanks, @jbenet. Good to see a lot of these comments!

**Comment by whyrusleeping** [comment]
@lgierth You're generally more eloquent than I am, could you handle adding better 'error case help text' here? https://github.com/ipfs/go-ipfs/pull/2939/files

**Comment by lgierth** [comment]
> @lgierth You're generally more eloquent than I am, could you handle adding better 'error case help text' here? https://github.com/ipfs/go-ipfs/pull/2939/files

Sure, here: #3158

**Comment by jbenet** [comment]
People have been checking boxes above without editing the post and adding:

```
- [ ] <problem description and issue links> --- resolved in <link-to-merged-PR>
```

see the "On this Code Review" rection

**Comment by jbenet** [comment]
Oh i see some RFCR things. i'll check those

**Comment by lgierth** [comment]
> People have been checking boxes above without editing the post and adding:

sorry that was sloppy -- I clarified that section about the gateway/blocking/reverting questions. Check the linked answers, I think these are resolved. (you're right, *I* shouldn't check them off, just provide you with enough info.)

**Comment by kevina** [comment]
I went through and clarified referenced pull requests with `---resolved in #XXX` for merged commits.  I also added `---needs review #XXX` for unmerged commits pending review.

Github makes it way to easy to accidentally toggle a checkbox and does not provide the history of edited commits to verify what I did.  Sorry, if I did this and caused something to be wrong.

**Comment by whyrusleeping** [comment]
@jbenet alright, your 'must haves' are all addressed at this point. For the rest, we will make issues out of each of them and get them addressed for 0.4.4

**Comment by Kubuxu** [comment]
Moved to 0.4.4 milestone.
